### PR TITLE
plan: return table dual when filter is false/null

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/7728

### What is changed and how it works?

Add false/null check after constant propagation

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: plan/rule_predicate_push_down.go [hunk 1]
```diff
@@ -34,6 +34,12 @@ func addSelection(p LogicalPlan, child LogicalPlan, conditions []expression.Expr
 		return
 	}
 	conditions = expression.PropagateConstant(p.context(), conditions)
+	// Return table dual when filter is constant false or null
```
Comment (XuHuaiyu) [inline plan/rule_predicate_push_down.go]: add a `.` at the end of this comment.

File: plan/rule_predicate_push_down.go [hunk 2]
```diff
@@ -55,6 +61,11 @@ func (p *LogicalSelection) PredicatePushDown(predicates []expression.Expression)
 	retConditions, child := p.children[0].PredicatePushDown(append(p.Conditions, predicates...))
 	if len(retConditions) > 0 {
 		p.Conditions = expression.PropagateConstant(p.ctx, retConditions)
+		// Return table dual when filter is constant false or null
```
Comment (XuHuaiyu) [inline plan/rule_predicate_push_down.go]: ditto

File: plan/rule_predicate_push_down.go [hunk 3]
```diff
@@ -375,3 +392,21 @@ func deriveOtherConditions(p *LogicalJoin, deriveLeft bool, deriveRight bool) (l
 	}
 	return
 }
+
+// conds2TableDual build a LogicalTableDual if cond is constant false or null
```
Comment (XuHuaiyu) [inline plan/rule_predicate_push_down.go]: s/ build/ builds

File: plan/rule_predicate_push_down.go [hunk 4]
```diff
@@ -375,3 +392,21 @@ func deriveOtherConditions(p *LogicalJoin, deriveLeft bool, deriveRight bool) (l
...
+func conds2TableDual(p LogicalPlan, conds []expression.Expression) LogicalPlan {
+	if len(conds) != 1 {
+		return nil
+	}
+	con, ok := conds[0].(*expression.Constant)
+	if !ok {
+		return nil
+	}
+	sc := p.context().GetSessionVars().StmtCtx
+	if isTrue, err := con.Value.ToBool(sc); (err == nil && isTrue == 0) || con.Value.IsNull() {
```
Comment (XuHuaiyu) [inline plan/rule_predicate_push_down.go]: Will it be possible that `err != nil && con.Value.IsNull()`

Comment (zz-jason) [reply plan/rule_predicate_push_down.go]: It is possible, but the constant is a `NULL` value, the `err` should be ignored.

**Pull request merged**